XLSX Pivot Table Downloads have 'Native' Pivot Table

adam-james-v opened the pull request:
closes #2473

The xlsx downloads work fine for regular tables. However, on Pivot Table downloads, the resulting file isn't pivoted and is formatted somewhat strangely (it contains additional rows related to totals, has an extra 'pivot-grouping' column).

Now, the downloaded xlsx file contains 2 sheets:
- the pivot sheet containing a 'native' Pivot table whose reference points to the data on the data sheet, and whose columns, rows, and values align with the cols, rows, and measures set up in Metabase
- the data sheet containing the question's rows unpivoted. If you were constructing a pivot table yourself, you would probably start with this shape of data.

I have Numbers locally, and things work correctly there:
<img width="1511" alt="image" src="https://github.com/metabase/metabase/assets/21064735/511f615d-abfb-415f-a6df-865555fa038e">

Uploading the same file to Google Sheets also produces a working Pivot Table:
<img width="1366" alt="image" src="https://github.com/metabase/metabase/assets/21064735/491bab16-06b9-48de-adcf-49286cf8f4fb">


And a video demonstrating interactivity with the native pivot:

https://github.com/metabase/metabase/assets/21064735/36cfffea-5808-4182-8701-06143b2d9046

adam-james-v committed:
XLSX Pivot Table Downloads have 'Native' Pivot Table

Addresses: #2473

The xlsx downloads work fine for regular tables. However, on Pivot Table downloads, the resulting file isn't pivoted
and is formatted somewhat strangely (it contains additional rows related to totals, has an extra 'pivot-grouping'
column).

Now, the downloaded xlsx file contains 2 sheets:
- the data sheet containing the question's rows unpivoted. If you were constructing a pivot table yourself, you would
probably start with this shape of data.
- the pivot sheet containing a 'native' Pivot table whose reference points to the data on the data sheet, and whose
columns, rows, and values align with the cols, rows, and measures set up in Metabase

adam-james-v committed:
Silly typo!

adam-james-v committed:
Exported Pivot uses correct aggregation (sum, count, stddev, avg, min, max)

adam-james-v committed:
Merge branch 'master' into excel-pivot-export

adam-james-v committed:
data for pivot exports now also uses the formatting

adam-james-v committed:
Add a test asserting that a pivot table exists in the xlsx file

adam-james-v committed:
add-row! is a method dispatching on sheet class

adam-james-v committed:
Merge branch 'master' into excel-pivot-export

adam-james-v committed:
remove unnecessary hint

adam-james-v committed:
Merge branch 'master' into excel-pivot-export

adam-james-v committed:
Merge branch 'master' into excel-pivot-export

escherize commented on src/metabase/query_processor/streaming/xlsx.clj:
```diff
@@ -447,38 +478,143 @@
+                                                   (dec idx)
+                                                   idx)) indices))
+        ordered-cols                    (vec (m/remove-nth pivot-grouping-key ordered-cols))
+        {:keys [pivot-rows
+                pivot-cols
+                pivot-measures
+                aggregation-functions]} (-> pivot-spec
+                                            (update :pivot-rows idx-shift)
+                                            (update :pivot-cols idx-shift)
+                                            (update :pivot-measures idx-shift)
+                                            (update :aggregation-functions #(vec (m/remove-nth pivot-grouping-key %))))
```
the destructuring feels a bit unneccesary here, what do you think of doing

```suggestion
       pivot-rows (idx-shift (:pivot-rows pivot-spec))
       ...
```

instead?

escherize commented on src/metabase/query_processor/streaming/xlsx.clj:
```diff
@@ -447,38 +478,143 @@
+
+;; Possible Functions: https://poi.apache.org/apidocs/dev/org/apache/poi/ss/usermodel/DataConsolidateFunction.html
+;; I'm only including the keys that seem to work for our Pivot Tables as of 2024-06-06
+(def ^:private agg-fn-k->pivot-agg-fn-enum
+  {:sum    DataConsolidateFunction/SUM
+   :avg    DataConsolidateFunction/AVERAGE
+   :min    DataConsolidateFunction/MIN
+   :max    DataConsolidateFunction/MAX
+   :count  DataConsolidateFunction/COUNT
+   :stddev DataConsolidateFunction/STD_DEV})
```
these keys look familiar, should this be defined together with `col->aggregation-function-key`? Do they need to be kept in sync?

adam-james-v replied:
I've consolidated this and the other fn into just one that handles this inside the xlsx namespace.

escherize commented on src/metabase/query_processor/streaming/xlsx.clj:
```diff
@@ -447,38 +478,143 @@
+                                                           ^AreaReference area-ref
+                                                           (CellReference. "A1")
+                                                           ^XSSFSheet data-sheet)]
+    (doseq [idx pivot-rows]
+      (.addRowLabel pivot-table idx))
+    (doseq [idx pivot-cols]
+      (.addColLabel pivot-table idx))
+    (doseq [idx pivot-measures]
+      (.addColumnLabel pivot-table (agg-fn-k->pivot-agg-fn-enum (get aggregation-functions idx DataConsolidateFunction/COUNT)) idx))
```
I noticed there weren't any added unit tests, and I think that would be a lot easier than recreating a universe when we need to test certain things

adam-james-v replied:
While I agree with this idea, the native-pivot function already takes in some data structures, the rows, the pivot-specifications, and some column data. 

From those pieces, it creates the workbook object, sheets, and creates the cell formatting objects too. It's a very imperative-java heavy section of code already, and with the existing xlsx functions in the rest of the namespace (the functions responsible for setting up cell styling and adding rows), I think this function is already taking in a reasonble 'data only' input, and dealing with things in an acceptable way, at least before this namespace is improved more hollistically.

escherize reviewed (changes requested):
Nice stuff. I had a few questions about this approach, don't take it as you need to take all of these suggestions, but I'd like to at least hear what you think about those questions.

adam-james-v committed:
Merge branch 'master' into excel-pivot-export

adam-james-v committed:
address review feedback:

 - consolidate the aggregation function key fns into one function in the .xlsx namespace
   this was moved out of postprocessing so that the xlsx specific fns don't litter the post processing namespace at
   this time

 - clean up an unnecessary destructiring inside the native-pivot function
 - don't shadow 'name' inside the body of a fn

adam-james-v committed:
Add tests that confirm zero-col and zero-row situations work

adam-james-v committed:
Merge branch 'master' into excel-pivot-export

escherize reviewed (approved):
Looks good!

I might try to combine those 3 atoms into one, but if you don't think it's a good idea, don't feel pressured.

adam-james-v committed:
Consolidate the atoms used to store pivot rows/options

iethree commented:
@adam-james-v "Addresses" is lovely english, but please use github [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) 🙏
